File: Day2_ControlFlow_Go/hard/hard_02.go
// FILE: hard_02.go
// TOPIC: Control Flow Deep Dive - Day 2 (Go)
// DIFFICULTY: Hard
//
// TASK: Cleanup / Unwinding Pattern: goto vs defer in Go
// -------------------------------------------------------
// In C / C++, developers often use `goto cleanup_stageX` to unwind resources
// in reverse order.
//
// In Go, `goto` is legal, but Go provides the `defer` keyword, which guarantees
// execution in LIFO (last-in, first-out) order when a function exits!
//
// TASK:
// 1. Write `initializeServiceGoto(fail1, fail2, fail3 bool) bool` using Go's `goto`
//    to implement the exact Linux-style cleanup unwind.
//
// 2. Test in main() with fail2 = true.
//
// EXPECTED OUTPUT:
//   Stage 1 Acquired (Network Buffer)
//   Failed Stage 2
//   Releasing Stage 1 (Network Buffer)
//   Initialization aborted.

package main

import "fmt"

// TODO: Write initializeServiceGoto here.

func main() {
	// TODO: Write your solution here.

}

File: Day2_ControlFlow/hard/hard_02.cpp
/*
 * FILE: hard_02.cpp
 * TOPIC: Control Flow Deep Dive - Day 2
 * DIFFICULTY: Hard
 *
 * TASK: Error Handling / Cleanup Pattern via goto (Linux Kernel Pattern)
 * -----------------------------------------------------------------------
 * In low-level C and C-style systems programming without RAII/exceptions,
 * allocating multiple resources sequentially often uses the "goto error_cleanup"
 * pattern to avoid code duplication.
 *
 * SCENARIO:
 * A system service needs 3 stages to initialize:
 *   Stage 1: Allocate Network Buffer
 *   Stage 2: Connect to Database
 *   Stage 3: Authenticate User
 *
 * If Stage 2 fails, we MUST release Stage 1 before exiting!
 * If Stage 3 fails, we MUST release Stage 2 AND Stage 1!
 *
 * 1. Write a function:
 *      bool initializeService(bool failStage1, bool failStage2, bool failStage3)
 *
 * 2. Implementation:
 *      - Stage 1:
 *          if (failStage1) { print "Failed Stage 1"; goto cleanup_none; }
 *          print "Stage 1 Acquired (Network Buffer)"
 *
 *      - Stage 2:
 *          if (failStage2) { print "Failed Stage 2"; goto cleanup_stage1; }
 *          print "Stage 2 Acquired (DB Connection)"
 *
 *      - Stage 3:
 *          if (failStage3) { print "Failed Stage 3"; goto cleanup_stage2; }
 *          print "Stage 3 Acquired (User Auth)"
 *
 *      print "Service Started Successfully!"
 *      return true;
 *
 *      // Stacked cleanups (unwind in reverse order):
 *      cleanup_stage2:
 *          print "Releasing Stage 2 (DB Connection)"
 *      cleanup_stage1:
 *          print "Releasing Stage 1 (Network Buffer)"
 *      cleanup_none:
 *          print "Initialization aborted."
 *          return false;
 *
 * 3. In main(), test the scenario where Stage 2 fails:
 *      initializeService(false, true, false);
 *
 * EXPECTED OUTPUT:
 *   Stage 1 Acquired (Network Buffer)
 *   Failed Stage 2
 *   Releasing Stage 1 (Network Buffer)
 *   Initialization aborted.
 */

#include <iostream>
using namespace std;

// TODO: Write initializeService here.

int main() {
    // TODO: Write your solution here.



    return 0;
}
